src/core/inductive.ml

```diff
@@ -0,0 +1,245 @@
+  let case_of : sym -> tbox = fun cons_sym ->
+    let codom : term list -> tvar list -> tbox -> tbox = fun ts xs _ ->
+      prf_of_p (List.map lift ts) (fapp cons_sym (List.rev_map _Vari xs))
+    in
+    let inj_var : tvar list -> tvar -> tvar = fun _ x -> x in
+    let build_rec_hyp : term list -> tvar -> tbox = fun ts x ->
+      prf_of_p (List.map lift ts) (_Vari x)
+    in
+    let domrec : term -> tvar -> tbox -> tbox -> tbox =
+      fun a x rec_hyp next ->
+      _Prod (lift a) (Bindlib.bind_var x (_Impl rec_hyp next))
+    in
+    let dom : term -> tvar -> tbox -> tbox = fun a x next ->
+      _Prod (lift a) (Bindlib.bind_var x next)
+    in
+    let f_rec : tbox -> tvar -> tbox -> tbox = fun a _ _ -> a in
+    let f_not_rec : tbox -> tvar -> tbox = fun a _ -> a in
+    let acc : tbox = _Type in
+    fold_cons_typ pos codom inj_var build_rec_hyp domrec dom ind_sym
+      cons_sym f_rec f_not_rec acc x_str
+  in
+
+  (* STEP 3: Create the conclusion of the inductive principle *)
+  let codom ts =
+    let x = Bindlib.new_var mkfree x_str in
+    let t = Bindlib.bind_var x (prf_of_p ts (_Vari x)) in
+    _Prod (fapp ind_sym ts) t
+  in
+  let conclusion = gen_ind_typ_codom pos ind_sym codom x_str in
+
+  (* STEP 4: Create the induction principle *)
+  let t =
+    List.fold_right (fun a b -> _Impl (case_of a) b) cons_list conclusion
+  in
+  let t = _Prod p_type (Bindlib.bind_var p t) in
+  if !log_enabled then
+    log_ind "The induction principle of the inductive type [%a] is %a"
+      Pretty.pp_ident (Pos.none ("ind_"^ind_sym.sym_name))
+      Print.pp_term (Bindlib.unbox t);
+  Bindlib.unbox t
+
+(** [gen_rec_rules ind_sym rec_sym cons_list] returns the p_rules associated
+    with an induction principle [rec_sym] of the inductive type [ind_sym]
+    (with its constructors [cons_list]). That means the command
+    inductive T : Π(i1:A1),...,Π(im:Am), TYPE := c1:T1 | ... | cn:Tn generates
+    a rule for each constructor. If Ti = Π(b1:B1), ... , Π(bk:Bk), T then the
+    rule for ci is:
+    rule ind_T p pic1 ... picn _ ... _ (ci x0 ... x(k-1)) -->
+    pici x0 t0? ... x(k-1) t(k-1)?
+    with m underscores and tj? = nothing if xj isn't a value of the inductive
+    type and tj? = (ind_T p pic1 ... picn _ ... _ xj) if xj is a value of the
+    inductive type T (i.e. xj = T v1 ... vm) *)
+let gen_rec_rules : sym -> sym -> sym list -> p_rule list =
+  fun ind_sym rec_sym cons_list ->
+
+  (* STEP 1: Create the common head of the rules *)
+  let rec_iden = P.iden rec_sym.sym_name in
+  let p_patt = P.patt0 "p" in
+  let add_arg t s = P.appl t (P.patt0 ("pi" ^ s.sym_name)) in
+  let head = P.appl rec_iden p_patt in
+  let common_head = List.fold_left add_arg head cons_list in
+
+  (* STEP 2: Create each p_rule according to a constructor *)
+  (* [gen_rule_cons cons_sym] creates a p_rule according to a constructor
+     [cons_sym]. *)
+  let gen_rule_cons : sym -> p_rule = fun cons_sym ->
+    let pos : popt = None in
```

Comment:
Thanks to use the position of the inductive type.

Reply:
Done